### ISSUE TITLE
Preselecting HTML render options in an remotely-sourced (AJAX) 

### ISSUE THREAD

**Issue by Toso82** [select2/select2#5210]
I don't find  how preselecting HTML render options in an remotely-sourced (AJAX). I read documentation https://select2.org/programmatic-control/add-select-clear-items, but not example of this.
I have a select2 latest version with my own templateSelection and templateResult.  So how can add initial data to my component.

**Comment by Sefrid** [comment]
参考 #5095

**Comment by Toso82** [comment]
I read this but i'm not so clear your solutions. Can I have a working example for this?. I don't undestand very well how i can add for use dataAdapter in select2

**Comment by Sefrid** [comment]
the key of solve question is use it's inner method "Select2.dataAdapter.convertToOptions(yourCustomData)" ,  then we can get a new options object that owner your whole data, next we still use it's inner method "Select2.dataAdapter.addOptions(lastStepWeGot)". well we insert options into select2 option. 
example:
`
		clearAll: function () {
			var _root = (this),
				settings = (_root.settings),
				element = (_root.element);

			$(element).val(null).trigger('change');
			_root.Select2.$selection.find('.select2-selection__rendered').prop('title', settings.placeholder);
			_root.Select2.dataAdapter.destroy();
			$(element).empty();
		},
		loadData: function (data, isClear, defaultValue, changeParam1) {
			var _root = (this),
				settings = (_root.settings),
				element = (_root.element),
				$option;

			if (isClear) {
				_root.clearAll();
			}
			$option = _root.Select2.dataAdapter.convertToOptions(data);
			_root.Select2.dataAdapter.addOptions($option);
			if (typeof defaultValue !== 'undefined') {
				if (isClear && (defaultValue == null)) {
					_root.Select2.$selection.find('.select2-selection__rendered').prop('title', settings.placeholder);
				}
				$(element).val(defaultValue).trigger('change', changeParam1);
			}
		},`

every times we  get data by ajax then use "loadData" method to insert options.

**Comment by maikelohcfg** [comment]
I found a more clean way, maybe not standard as should be, but i tested and work properly. 
I used a select2 with ajax data source, but need set a selected value with extra data fields wich are used on other methods and events on my page.

Official [doc](https://select2.org/programmatic-control/add-select-clear-items) say you must do this
```
 var option = new Option(data.full_name, data.id, true, true);
    studentSelect.append(option).trigger('change');
    studentSelect.trigger({
        type: 'select2:select',
        params: {
            data: data
        }
    });
```
this is fine but the trick is this

```
var data    = {
     id : 1,
     full_name: "Student 1",
    group: "Group #1",
    anotherField: "With some value"
}
var option = $(new Option(data.full_name, data.id, true, true));
option.data('data',
                    {
                        id: data.id,
                        text: data.full_name),
                        data: data,
                        disabled: option.prop('disabled'),
                        selected: option.prop('selected'),
                        title: option.prop('title'),
                        element : option[0]
                    }
                )
studentSelect.append(option).trigger('change');
studentSelect.trigger({
        type: 'select2:select',
        params: {
            data: data
        }
});
```
you should trigger "select2:select" if you already have bind some listener to this event, so it will receive the new data, try also 

`studentSelect.select2('data')`

and you will get the field "data" with all the fields of student.

Remember this solution is when you already have instantiated the select2 and need select a custom value without re instantiate the component. If you need at instantiate moment of ajax select2 set default selected values, you must use the "data" config parameter as say the official doc

**Comment by TCook7** [comment]
Thanks @maikelohcfg 
I did a slight variation to allow the existing `templateSelection` function I had to work both with this as well as the AJAX data without any code changes.

```
let option = new Option(data.formattedName, data.id, true, true);
$(option).data('data', data);
contributorSelect.append(option);
```

I didn't need to do the trigger on select2:select for my situation.
I also noticed the `.trigger('change')` seemed redundant. It looks like doing the `.append(option)` already triggers a change.

It would be great if this was added to the Programmatic Control section on the Select2 site. 
[](https://select2.org/programmatic-control/add-select-clear-items)
